### ISSUE TITLE
New File/Folder picker for Reunion

### ISSUE THREAD

**Issue by ZenBird** [microsoft/ProjectReunion#88]
# Proposal: New File/Folder picker for Reunion

## Summary
This proposal introduces an attempt to design a new UWP File/Folder picker for Reunion that overcomes the limitations of the UWP File/Folder picker currently available in the SDK

## Rationale

Listed below are the limitations of the current File/Folder picker
* Unable to pick files and folders in a single instance of the picker dialog
* Unable to pick multiple folders

This proposal attempts to address the above issues for down-level OS'es (RS1 and higher).

## Scope

| Capability | Priority |
| :---------- | :------- |
| This proposal will allow developers to pick files and folders in a single instance of the picker dialog | Must |
| This proposal will allow developers to pick multiple folders | Must |
| This proposal will allow developers to specify paths that are outside the application install path (with permissions) | Must |
| This proposal will allow applications to pick Files/Folders on behalf of a user | Could |
| This proposal will allow applications to customize the File/Folder dialog (within limitations) | Should |

## Open Questions

* Will applications need special permissions to use the file/folder picker other than storage permissions.

**Comment by Felix-Dev** [comment]
It would also be helpful to enable developers to freely set the picker's start location. Currently, these pickers have a [SuggestedStartLocation](https://docs.microsoft.com/en-us/uwp/api/windows.storage.pickers.fileopenpicker.suggestedstartlocation?view=winrt-19041#Windows_Storage_Pickers_FileOpenPicker_SuggestedStartLocation) API which only lets the developer pick from a small set of (common) file system locations (such as the user's desktop or download folder). This API could be changed to accept any file/folder path the developer wants to specify, for example like this:
```cs
public string SuggestedStartLocation { get; set; }
```
to be used like:
```cs
var filePicker = new FileOpenPicker()
{
    SuggestedStartLocation = "MyCustomFolderFilePath"
}
```
Developers would still be able to easily set the start location to the folders they currently can pick for the `SuggestedStartLocation` like this:
```cs
var filePicker = new FileOpenPicker()
{
    SuggestedStartLocation = Environment.GetFolderPath(Environment.SpecialFolder.MyPictures)
};
```
This would also help mitigate the UX issue being called out in issue #45.

**Comment by soumyamahunt** [comment]
Also FilePicker doesn't allow apps to add custom options like it can be done with open and save as dialog boxes in case of win32 apps:

![WINWORD_ZEjbI5QRXd](https://user-images.githubusercontent.com/51203900/85509465-b1b0a880-b613-11ea-8ee8-34da25e75dc8.png)

![notepad_hoTBxpTwTD](https://user-images.githubusercontent.com/51203900/85509469-b2e1d580-b613-11ea-8009-b0e36f0f501e.png)

Being able to add custom options like in can be done with print dialog will also be helpful.

**Comment by harvinders** [comment]
Also consider defining proper AutomationIds, currently two properties (ComboBox and TextBox in ComboBox) have automation id of 1148, in the OS file picker. This makes UI testing tricky.

**Comment by mdtauk** [comment]
Not sure if this gets implemented in the OS, as a Reunion UI/framework - or if it comes with the forthcoming Windows 10X Files app - but these old UI surfaces, should be replaced with Xaml UIs - for apps compiled using Reunion APIs

**Comment by Ben3094** [comment]
Please add other applications able to provide "file" as the "Camera" app in Windows 10

**Comment by driver1998** [comment]
WinRT apps support picking files from other app, and on Windows 10, it is implemented like a hack, with XAML UI showing in a cutout of the Old Explorer based file picker.

One protential usage would be like calling Camera app to take selfees when uploading profile pictures.

This feature is currently not available in Win32 Apps.

The new file picker should be fully WinUI based, and add support for this in Win32 Apps. It could be done with a path to a temp file, and automatically deleted after the file is closed by the app.

**Comment by mdtauk** [comment]
Just linking to this WinUI topic, where we are discussing building a new Reunion WinUI File Picker Dialog for this API proposal
https://github.com/microsoft/microsoft-ui-xaml/issues/2854

![dialog mockup](https://user-images.githubusercontent.com/7389110/86991800-dec69480-c197-11ea-8491-3e62bcd9d389.png)

**Comment by PeterSmithRedmond** [comment]
I would love a folder picker where as a developer I can suggest a OneDrive folder. Right now I can provide a suggestion that the folder picker start in the My Documents folder, but I can't then recommend that it start in an actual useful sub-folder like 'My Documents\Roaming_Bookmarks'

Here's my scenario: I want people to be able to pick a folder that will be used to synchronize some bookmark data. Each computer they run on, they have to pick the same folder (this assumes that they want their bookmarks to roam; if they don't, the default is to just store them locally). But users are terrible at picking a folder!

1. By default, I can point them to the My Documents folders, but this clutters up the Documents folder.
2.  I can pop up a dialog and suggest to users that they pick a folder on OneDrive, but I can't set what that folder is
3. When they do the same thing on a new computer, they have to remember which folder they picked before, and that's hard

TL/DR: save folder picker should let me as a developer specify a currently non-existing OneDrive folder by name that will be created if the user agree

**Comment by mevey** [comment]
@ZenBird  Any updates on this based on the feedback thus far?

**Comment by mdtauk** [comment]
Can we get some indication where this proposal stands?

Is it being actively considered or worked on?

What version of Reunion is this likely to land in?

**Comment by Jaiganeshkumaran** [comment]
> WinRT apps support picking files from other app, and on Windows 10, it is implemented like a hack, with XAML UI showing in a cutout of the Old Explorer based file picker.
> 
> One protential usage would be like calling Camera app to take selfees when uploading profile pictures.
> 
> This feature is currently not available in Win32 Apps.
> 
> The new file picker should be fully WinUI based, and add support for this in Win32 Apps. It could be done with a path to a temp file, and automatically deleted after the file is closed by the app.

Unlike Windows 8, both file pickers in UWP and Win32 almost look the same except a few minor differences. File Picker contract is powerful however you don't get them when you are working with Win32 apps which shouldn't be the case in the first place. However the WinRT file picker returns a StorageFile which not only can represents a file on the filesystem but also something provided by another app. The old File Picker API however returns the path and I'm not sure how would that work with these apps  if they aren't stored on the local file system. Anyway the old API could become a wrapper on top of FileOpenPicker and return the path property on it. Another thing is there are even older file pickers which sometimes show up in certain apps from pre-Vista era which needs to be replaced with a modern one.

**Comment by bogdan-patraucean** [comment]
Just came across this and have the same problem, not even attaching the window handle works.

**Comment by hawkerm** [comment]
Currently also the `FileOpenPicker` only has [`FileTypeFilter`](https://learn.microsoft.com/en-us/uwp/api/windows.storage.pickers.fileopenpicker.filetypefilter) which doesn't let you name the type of file displayed in the dialog (only list extensions), see this Stack Overflow question from years ago here: https://stackoverflow.com/questions/32201831/how-to-properly-specify-fileopenpicker-filetypefilter-syntax-for-uwp

Which is weird, as the `FileSavePicker` takes a map to enable this scenario with [`FileTypeChoices`](https://learn.microsoft.com/en-us/uwp/api/windows.storage.pickers.filesavepicker.filetypechoices).

The `FileOpenPicker` should also just use the same map as `FileTypeChoices` as well so that file type names (along with those with multiple extensions like images can be grouped).

**Comment by castorix** [comment]
> Listed below are the limitations of the current File/Folder picker
> Unable to pick files and folders in a single instance of the picker dialog
> Unable to pick multiple folders

[IFileDialogCustomize](https://learn.microsoft.com/en-us/windows/win32/api/shobjidl_core/nn-shobjidl_core-ifiledialogcustomize) does that

**Comment by Tam66662** [comment]
So it's only been _three and a half years_ since this request was made. The usability of the FileOpenPicker and FolderPicker is practically zero. I have a UI with three FileOpenPickers from three distinct folders that never change. But I literally have to _re-select_ each folder every time the dialog pops up because it doesn't have a start location. How do you expect developers to adopt WinUI for web and mobile application development if the UX is like this? Please please please fix these controls.

**Comment by tpoint75** [comment]
Its unbelievable, that we can't set a custom start location in the FileOpenPicker. What a shame!

**Comment by BendicoE** [comment]
Being unable to set the start location for the FileOpenPicker and FileSavePicker in WinUI 3 is a HUGE drawback. Basically, we are unable to help the user pick the places where he has recently been opening or saving files. All modern Windows application has this, and our application cannot provide it. This is a MUST HAVE in future versions of the Windows App SDK!!!

**Comment by JaiganeshKumaran** [comment]
> Basically, we are unable to help the user pick the places where he has recently been opening or saving files.
I believe you can achieve this by setting the [SettingsIdentifier property](https://learn.microsoft.com/en-us/uwp/api/windows.storage.pickers.fileopenpicker.settingsidentifier?view=winrt-2262).

**Comment by BendicoE** [comment]
@JaiganeshKumaran I believe this only works within an application session. Our application persists recently used files and places between sessions and even after rebooting the system. Or am I misunderstanding something here?

**Comment by chanallenk** [comment]
The existing folder picker is also a huge flaw in design. I have hundreds of users who are constantly confusing a folderdialog with a filedialog, expecting them to be able to see their files present when browsing for a folder. When the files are missing (because this is a folder dialog), they think the app is broken or not detecting files they should be seeing. The[ windows forms folder dialog](https://i.stack.imgur.com/Y5zXQ.png) was much more intuitive

**Comment by castorix** [comment]
> The[ windows forms folder dialog](https://i.stack.imgur.com/Y5zXQ.png) was much more intuitive

This is [SHBrowseForFolder](https://learn.microsoft.com/en-us/windows/win32/api/shlobj_core/nf-shlobj_core-shbrowseforfolderw), that you can call in WinUI

**Comment by castorix** [comment]
> I was curious if it would be possible to implement "Setting default location for a file picker" (see [dotnet/maui#9212 (comment)](https://github.com/dotnet/maui/issues/9212#issuecomment-1826125726)) in MAUI ... but it seems hard.
> 
> Has anyone come up with a workaround to do it?

On Windows, default folder can be set with [IFileDialog::SetFolder](https://learn.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-ifiledialog-setfolder) or IFileDialog::SetDefaultFolder
(I used it in the test I had done  ([WinUI3_IFileDialog](https://github.com/castorix/WinUI3_IFileDialog)) as workaround...)

**Comment by dotMorten** [comment]
Looks like this is 1.8-preview1: https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/preview-channel#new-apis

**Comment by ssparach** [comment]
Thank you for raising this issue.

New [StoragePicker APIs ](https://learn.microsoft.com/windows/apps/windows-app-sdk/stable-channel#storage-pickers)have been introduced in WindowsAppSDK version 1.8 (Stable). We’re resolving this issue as completed.

For any further feedback or suggestions regarding the APIs, please feel free to open a new issue. Thanks again for the engagement!
